Plugins/NativeXr/Source/NativeXr.cpp

```diff
@@ -609,68 +623,99 @@ namespace Babylon
-                    m_sessionState->GraphicsImpl.AddTexture(depthTexture, static_cast<uint16_t>(view.DepthTextureSize.Width), static_cast<uint16_t>(view.DepthTextureSize.Height), false, 1, depthTextureFormat);
-
-                    arcana::make_task(m_sessionState->GraphicsImpl.AfterRenderScheduler(), arcana::cancellation::none(), [colorTexture, depthTexture, &view]() {
-                        bgfx::overrideInternal(colorTexture, reinterpret_cast<uintptr_t>(view.ColorTexturePointer));
-                        bgfx::overrideInternal(depthTexture, reinterpret_cast<uintptr_t>(view.DepthTexturePointer));
-                    }).then(m_runtimeScheduler, m_sessionState->CancellationSource, [this, thisRef{shared_from_this()}, colorTexture, depthTexture, &view]() {
-                        std::array<bgfx::Attachment, 2> attachments{};
-                        attachments[0].init(colorTexture);
-                        attachments[1].init(depthTexture);
-                        auto frameBufferHandle = bgfx::createFrameBuffer(static_cast<uint8_t>(attachments.size()), attachments.data(), false);
-
-                        auto& frameBuffer{m_sessionState->GraphicsImpl.AddFrameBuffer(frameBufferHandle,
-                            static_cast<uint16_t>(view.ColorTextureSize.Width),
-                            static_cast<uint16_t>(view.ColorTextureSize.Height),
-                            true)};
-
-                        // WebXR, at least in its current implementation, specifies an implicit default clear to black.
-                        // https://immersive-web.github.io/webxr/#xrwebgllayer-interface
-                        frameBuffer.Clear(m_sessionState->GraphicsImpl.GetUpdateToken().GetEncoder(), BGFX_CLEAR_COLOR | BGFX_CLEAR_DEPTH | BGFX_CLEAR_STENCIL, 0, 1.0f, 0);
-
-                        m_sessionState->TextureToFrameBufferMap[view.ColorTexturePointer] = &frameBuffer;
-
-                        auto jsWidth{Napi::Value::From(m_env, view.ColorTextureSize.Width)};
-                        auto jsHeight{Napi::Value::From(m_env, view.ColorTextureSize.Height)};
-                        auto jsFrameBuffer{Napi::External<FrameBuffer>::New(m_env, &frameBuffer)};
-                        m_sessionState->FrameBufferToJsTextureMap[&frameBuffer] = Napi::Persistent(m_sessionState->CreateRenderTexture.Call({jsWidth, jsHeight, jsFrameBuffer}).As<Napi::Object>());
+                    auto depthTexture = bgfx::createTexture2D(textureWidth, textureHeight, false, textureLayers, depthTextureFormat, BGFX_TEXTURE_RT);
+                    m_sessionState->GraphicsImpl.AddTexture(depthTexture, textureWidth, textureHeight, false, textureLayers, depthTextureFormat);
+
+                    arcana::make_task(m_sessionState->GraphicsImpl.AfterRenderScheduler(), arcana::cancellation::none(), [colorTexture, depthTexture, &viewConfig]() {
```

Comment:
>viewConfig [](http://example.com/codeflow?start=151&length=10)

i might be over thinking this, but if view config is a reference declared here in this function scope, is there any way to guarantee that it still exists when this task executes? should we just be feeding in a copy of the reference compared to a reference to the reference?

Reply:
The capture syntax here implies a reference to the original memory, in this case an entry in `m_sessionState->TextureToViewConfigurationMap`. As long as this lambda executes while the session is still alive (and the map hasn't been cleared), this will work.

So this will work as long as nothing super weird happens, but I could potentially see something happening where the session is ended before the textures get overridden, so maybe as an extra measure I could through the session state cancellation source into the first task as well. @syntheticmagus might have more thoughts.